### PR TITLE
add plan route

### PR DESCRIPTION
**Pull request by alisa911**
Screenshot 2022-07-04 at 21 24 51

### CODE REVIEW

File: map/src/context/AppContext.js [hunk 1]
```diff
@@ -327,7 +333,29 @@ export const AppContextProvider = (props) => {
...
+        file: null,
+        groups: [],
+        groupsUnique: [],
+        readFirst: false,
+        visibleMarker: {
+            prev: null,
+            current: null
+        }
+    });
+    const [planRoute, setPlanRoute] = useState(false);
```
Comment (vshcherb) [inline map/src/context/AppContext.js:346]: what's difference between planRoute / newRoute / editRoute? sounds as same object

File: map/src/setupProxy.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
 const { createProxyMiddleware } = require('http-proxy-middleware');
 module.exports = function (app) {
+    // let proxy = createProxyMiddleware({
```
Comment (vshcherb) [inline map/src/setupProxy.js:3]: probably not needed

File: map/src/context/AppContext.js [hunk 3]
```diff
@@ -327,7 +333,29 @@ export const AppContextProvider = (props) => {
...
+    const [newRoute, setNewRoute] = useState({
+        newRouteLayer: new Layer(),
+        pointsList: [],
+        finished: false,
+        distance: 0
+    });
+    const [editor, setEditor] = useState({
```
[the remark this answers is not in the document]
Comment (alisa911) [reply map/src/context/AppContext.js:353]: trackEditor would be better?

File: map/src/context/AppContext.js (hunk 1 above)
Comment (alisa911) [reply map/src/context/AppContext.js:346]: planRoute - flag for mode when panelButton appear
newRoute - object
editor - object with edit functions

File: map/src/context/AppContext.js [hunk 4]
```diff
@@ -382,7 +410,10 @@ export const AppContextProvider = (props) => {
         weatherPoint, setWeatherPoint,
         routeTrackFile, setRouteTrackFile,
         searchCtx, setSearchCtx,
-        favorites, setFavorites
+        favorites, setFavorites,
+        planRoute, setPlanRoute,
```
Comment (vshcherb) [inline map/src/context/AppContext.js:414]: 2 objects needed
currentlyEditTrack - null or object { }
localEditedTracks - [ ] - array of local edited objects with flag on / of

currentlyEditTrack.prototype.deletePointMap(ind) = function () {};
currentlyEditTrack.prototype.deletePoint(ind) = function... ( ... this.deletePointMap(ind); ... )

currentlyEditTrack.prototype.clearMap(ind) = function () {};
currentlyEditTrack.prototype.clear() = function... ( ... this.clearMap(); ... )

File: map/src/components/layers/PlanRouteLayer.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                type: "application/gpx+xml",
+            });
+            let gpxLayer = {};
+            gpxLayer.name = 'local:' + file.name;
+            gpxLayer.localContent = `${process.env.REACT_APP_GPX_API}/gpx/get-gpx-file?name=` + encodeURIComponent(file.name);
+            gpxLayer.local = true;
+            Utils.uploadFile(ctx.gpxFiles, ctx.setGpxFiles, ctx, gpxLayer, file);
+        }
+    }, [newFileName, setNewFileName]);
+
```
Comment (vshcherb) [inline map/src/components/layers/PlanRouteLayer.js:39]: useEffect( () => {
currentlyEditTrack.prototype.clearMap() = function() {
....
};
}, []);

File: map/src/context/AppContext.js [hunk 6]
```diff
@@ -310,24 +321,38 @@ export const AppContextProvider = (props) => {
...
+        groupsUnique: [],
+        readFirst: false,
+        visibleMarker: {
+            prev: null,
+            current: null
+        }
+    });
+    const [currentlyEditTrack, currentlyEditTrackDispatch] = useReducer(CurrentlyEditTrackReducer, null);
+    const [createdTracks, setCreatedTracks] = useState(localStorage.getItem('createdTracks') !== null ? JSON.parse(localStorage.getItem('createdTracks')) : []);
+    const [createNewTrack, setCreateNewTrack] = useState(false);
```
Comment (vshcherb) [inline map/src/context/AppContext.js:355]: use currentlyEditTrack instead of createNewTrack?

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        } else if (prevCreatedTracks && prevCreatedTracks.length >= ctx.createdTracks.length) {
+            deleteOldRoute(map);
+        }
+    }, [ctx.createdTracks, ctx.setCreatedTracks]);
+
+
+    useEffect(() => {
+        if (ctx.createNewTrack) {
+            if (ctx.currentlyEditTrack) {
+                deleteOldRoute(map);
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:77]: this is the only method with map and could be handled with useEffect(...state.clean ?)

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+export default function PlanRouteLayer() {
+    const ctx = useContext(AppContext);
+    const map = useMap();
+
+    const [openSaveDialog, setOpenSaveDialog] = useState(false);
+    const [openPanelButtons, setOpenPanelButtons] = useState(false);
+    const [prevCreatedTracks, setPrevCreatedTracks] = useState(null);
+
+    if (ctx.currentlyEditTrack) {
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:17]: useEffect ( ? map ? )

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        map.on("editable:vertex:dragend", e => {
+            ctx.currentlyEditTrackDispatch({
+                type: 'dragendClick',
+            })
+        });
+    }
+
+    useEffect(() => {
+        let selectedTrack = ctx.createdTracks.find(t => t.selected === true);
+        if (selectedTrack) {
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:41]: split by methods name (comment)

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+    useEffect(() => {
+        if (ctx.currentlyEditTrack) {
+            if (ctx.currentlyEditTrack.startDraw) {
+                if (!ctx.currentlyEditTrack.newRouteLayer) {
+                    ctx.currentlyEditTrackDispatch({
+                        type: 'start',
+                        newRouteLayer: map.editTools.startPolyline(),
+                    })
+                    ctx.setSelectedGpxFile(null);
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:110]:
```
  ctx.setSelectedGpxFile(null);
  ctx.setWeatherPoint(null);
```
Should be replaced with 1 method ctx.setContextMenuObject

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                }
+            }
+        );
+        layersWithPolyline.forEach(function (item) {
+            map.removeLayer(item);
+        });
+    }
+
+    useEffect(() => {
+        if (ctx.currentlyEditTrack) {
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:103]: split by methods or something - 3 unclear if blocks

File: map/src/contextmenu/components/tabs/PointsTab.jsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                </ListItemIcon>
+                <ListItemText inset>
+                    <Typography variant="inherit" noWrap>
+                        Point - {index + 1}<br/>
+                        {point.dist === 0 ? "start" : Math.trunc(point.dist) + " m"}
+                    </Typography>
+                </ListItemText>
+                <ListItemAvatar>
+                    <IconButton x={{mr: 1}} onClick={() => {
+                        ctx.currentlyEditTrackDispatch({
```
Comment (vshcherb) [inline map/src/contextmenu/components/tabs/PointsTab.jsx:36]: 2 dispatches? addDeletePointIndex -> deletePoint ?

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                ctx.currentlyEditTrack.pointsList.splice(ctx.currentlyEditTrack.deletePoint, 1);
+                deleteOldRoute(map);
+                ctx.currentlyEditTrack.pointsList = Utils.getPointsDist(ctx.currentlyEditTrack.pointsList);
+                ctx.currentlyEditTrack.newRouteLayer = map.editTools.addPolylineByPoints(ctx.currentlyEditTrack.pointsList);
+
+                ctx.currentlyEditTrackDispatch({
+                    type: 'deletePoint',
+                    pointsList: ctx.currentlyEditTrack.pointsList,
+                    layer: ctx.currentlyEditTrack.newRouteLayer,
+                    distance: ctx.currentlyEditTrack.pointsList[ctx.currentlyEditTrack.pointsList.length - 1].dist
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:141]: - Distance should be responsibility of Object and never calculated in other places

File: map/src/map/components/layers/PlanRouteLayer.js [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            //check delete point
+            if (ctx.currentlyEditTrack.deletePoint && ctx.currentlyEditTrack.deletePoint !== -1) {
+
+                ctx.currentlyEditTrack.pointsList.splice(ctx.currentlyEditTrack.deletePoint, 1);
+                deleteOldRoute(map);
+                ctx.currentlyEditTrack.pointsList = Utils.getPointsDist(ctx.currentlyEditTrack.pointsList);
+                ctx.currentlyEditTrack.newRouteLayer = map.editTools.addPolylineByPoints(ctx.currentlyEditTrack.pointsList);
+
+                ctx.currentlyEditTrackDispatch({
+                    type: 'deletePoint',
```
Comment (vshcherb) [inline map/src/map/components/layers/PlanRouteLayer.js:138]:
```
type: 'refreshLayer',
layer: ctx.currentlyEditTrack.newRouteLayer,
```

File: map/src/data/tracks/CurrentlyEditTrackReducer.js [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        case 'clean': {
+            return {
+                ...state,
+                newRouteLayer: null,
+                pointsList: []
+            };
+        }
+        case 'addDeletePointIndex': {
+            return {
+                ...state,
```
Comment (vshcherb) [inline map/src/data/tracks/CurrentlyEditTrackReducer.js:71]: Here 
ctx.currentlyEditTrack.pointsList.splice(ctx.currentlyEditTrack.deletePoint, 1);
 ctx.currentlyEditTrack.pointsList = Utils.getPointsDist(ctx.currentlyEditTrack.pointsList);

-> (replace deletePoint -> refreshLayer)
refreshLayer: true

File: map/src/contextmenu/components/MapContextMenu.jsx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        if (ctx.currentlyEditTrack && ctx.currentlyEditTrack.newRouteLayer && ctx.currentlyEditTrack.newRouteLayer._latlngs) {
+            return new EditTrackTabList().create(ctx, graphWidth);
+        }
+        if (ctx.weatherPoint?.day && ctx.weatherPoint?.week) {
+            return new WeatherTabList().create(ctx, graphWidth);
+        }
+    }
+
+    function closeContextMenu() {
+        ctx.setSelectedGpxFile(null);
```
Comment (vshcherb) [inline map/src/contextmenu/components/MapContextMenu.jsx:63]: setContextMenuObjectType(null) (possible options 'weather-point', 'create-track', 'selected-track', 'favorite')

File: map/src/data/tracks/editTrack/EditTrackReducer.js [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+                prepareMap: false,
+                stopDraw: false,
+                showTrack: true,
+                deleteLayer: false
+            };
+        }
+        case EditTrackAction.click: {
+            return {
+                ...state,
+                pointsList: getPointList(action.e),
```
Comment (vshcherb) [inline map/src/data/tracks/editTrack/EditTrackReducer.js:62]: getPointList - bad name (edit ?)

File: map/src/context/AppContext.js [hunk 18]
```diff
@@ -317,25 +323,61 @@ export const AppContextProvider = (props) => {
...
+        visibleMarker: {
+            prev: null,
+            current: null
+        }
+    });
+
+    const [createdTracks, setCreatedTracks] = useState(localStorage.getItem('createdTracks') !== null ? JSON.parse(localStorage.getItem('createdTracks')) : []);
+    const [contextMenuObjectType, setContextMenuObjectType] = useState(null);
+
+    useEffect(() => {
```
Comment (vshcherb) [inline map/src/context/AppContext.js:360]: Why is it needed?

File: map/src/drawer/components/favorite/FavoriteItem.jsx [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    },
+    item: {
+        '& li': {
+            minHeight: '50px !important',
+            maxHeight: '50px !important',
+        }
+    },
+    text: {
+        '& p': {
```
Comment (vshcherb) [inline map/src/drawer/components/favorite/FavoriteItem.jsx:26]: It looks strange cause we didn't use any custom style for other elements in drawer. Needs to be investigated.

File: map/src/drawer/components/tracks/CreatedTrackItem.jsx [hunk 20]
```diff
@@ -0,0 +1,67 @@
+import React, {useContext, useState} from "react";
+import AppContext from "../../../context/AppContext";
+import {ListItemText, MenuItem, Switch, Tooltip, Typography} from "@mui/material";
+import CreatedTrackUtils from "../../util/CreatedTrackUtils";
+import {BaseBuilder, buildGPX} from "gpx-builder";
+import {Metadata, Point} from "gpx-builder/dist/builder/BaseBuilder/models";
+
+export default function CreatedTrackItem({track, index}) {
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/CreatedTrackItem.jsx:8]: Bad Name

File: map/src/drawer/components/tracks/CreatedTrackItem.jsx [hunk 21]
```diff
@@ -0,0 +1,67 @@
+import React, {useContext, useState} from "react";
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/CreatedTrackItem.jsx:1]: This file should be a separate pull request to not block Favorites

File: map/src/drawer/components/tracks/VisibleTrackGroup.jsx [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+            </ListItemIcon>
+            <ListItemText> Visible </ListItemText>
+            <Typography variant="body2" color="textSecondary">
+                {getVisibleTracksLength() > 0 ? `${getVisibleTracksLength()}` : ''}
+            </Typography>
+            {(visibleTracks.created.length > 0 || visibleTracks.cloud.length > 0) ? <></> : visibleTracksOpen ?
+                <ExpandLess/> : <ExpandMore/>}
+        </MenuItem>
+        <Collapse in={visibleTracksOpen} timeout="auto" unmountOnExit>
+            {visibleTracks.created.length > 0 && visibleTracks.created.map((track, index) => {
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/VisibleTrackGroup.jsx:48]: This file should be a separate pull request to not block Favorites

File: map/src/util/Utils.js [hunk 23]
```diff
@@ -53,8 +52,101 @@ async function getFileData(file) {
...
+                let d = getDistance(list[index].lat, list[index].lng, list[index - 1].lat, list[index - 1].lng);
+                dist += d;
+                list[index].dist = dist;
+            }
+        }
+    }
+    return list;
+}
+
+function removeDuplicatesPoints(list) {
```
Comment (vshcherb) [inline map/src/util/Utils.js:88]: This is not needed function likely

File: map/src/drawer/components/tracks/TracksMenu.jsx [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        setTracksGroups(tracksGroups);
+
+    }, [ctx.listFiles, ctx.setListFiles]);
+
+
+    useEffect(() => {
+        if (ctx.gpxFiles) {
+            visibleTracks.cloud = [];
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/TracksMenu.jsx:60]: setVisibleTracks looks strange it should be no difference cloud or not cloud is visible.

File: map/src/drawer/components/tracks/TracksMenu.jsx [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+            Object.values(ctx.gpxFiles).forEach((f) => {
+                if (f.url) {
+                    visibleTracks.cloud.push(f);
+                }
+            })
+        }
+        setVisibleTracks({...visibleTracks});
+    }, [ctx.gpxFiles, ctx.setGpxFiles]);
+
+    useEffect(() => {
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/TracksMenu.jsx:70]: This should be separate pull request

File: map/src/drawer/components/tracks/TracksMenu.jsx [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+            })
+            if (defGroup) {
+                tracksGroups.splice(tracksGroups.indexOf(defGroup), 1);
+                tracksGroups.unshift(defGroup);
+            }
+        }
+
+        setTracksGroups(tracksGroups);
+
+    }, [ctx.listFiles, ctx.setListFiles]);
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/TracksMenu.jsx:55]: useEffect on setListFiles ?

File: map/src/drawer/util/LocalTrackUtils.js [hunk 27]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (vshcherb) [inline map/src/drawer/util/LocalTrackUtils.js:1]: Looks very small function and probably doesn't need separate module

File: map/src/map/layers/CreatedTrackLayer.js [hunk 28]
```diff
@@ -0,0 +1,59 @@
+import React, {useContext, useEffect, useState} from "react";
+import AppContext from "../../context/AppContext";
+import {useMap} from "react-leaflet";
+import L from "leaflet";
+import MarkerIcon from "../MarkerIcon";
+
+
+export default function CreatedTrackLayer() {
```
Comment (vshcherb) [inline map/src/map/layers/CreatedTrackLayer.js:8]: Not needed now

File: map/src/map/components/OsmAndMap.jsx [hunk 29]
```diff
@@ -71,9 +65,9 @@ const OsmAndMap = () => {
 
     return (
         <MapContainer center={position} zoom={5} className={classes.root} minZoom={1} maxZoom={20}
-                      zoomControl={false} whenReady={whenReadyHandler} contextmenu={true} contextmenuItems={[]}
+                      zoomControl={false} whenReady={whenReadyHandler} contextmenu={true} contextmenuItems={[]} editable={true}
         >
-
+            <CreatedTrackLayer/>
```
Comment (vshcherb) [inline map/src/map/components/OsmAndMap.jsx:70]: Not needed now

File: map/src/drawer/components/tracks/LocalTrackGroup.jsx [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+
+const useStyles = makeStyles({
+    button: {
+        maxWidth: '120px !important',
+        maxHeight: '40px',
+        minWidth: '120px !important',
+        minHeight: '40px'
+    },
+    group: {
+        '& li': {
```
Comment (vshcherb) [inline map/src/drawer/components/tracks/LocalTrackGroup.jsx:22]: Looks very similar to global style for other items if we need for all of them it's better to have it common

File: map/src/drawer/util/CreatedTrackUtils.js [hunk 31]
```diff
@@ -0,0 +1,25 @@
+function unselectedTrack(tracks, track) {
```
Comment (vshcherb) [inline map/src/drawer/util/CreatedTrackUtils.js:1]: Probably separate request is needed

File: map/src/drawer/components/tracks/LocalTrackGroup.jsx (hunk 30 above)
Comment (alisa911) [reply map/src/drawer/components/tracks/LocalTrackGroup.jsx:22]: This class used only for MenuItem. We want lines will be same height. We don't have do it other way.

File: map/src/contextmenu/components/tabs/GeneralInfoTab.jsx [hunk 32]
```diff
@@ -29,29 +29,29 @@ export default function GeneralInfo({ width, summary, url }) {
     if (info?.hasElevationData) {
         updownhill = "Uphill/downhill: " + info.diffElevationUp.toFixed(0)
             + "/" + info?.diffElevationDown.toFixed(0) + " m";
-        elevation = "Elevation (min/avg/max): " +
+        elevation = "ElevationTab (min/avg/max): " +
```
[the remark this answers is not in the document]
Comment (alisa911) [reply map/src/contextmenu/components/tabs/GeneralInfoTab.jsx:32]: +

File: map/src/drawer/components/favorite/FavoriteItem.jsx (hunk 19 above)
Comment (alisa911) [reply map/src/drawer/components/favorite/FavoriteItem.jsx:26]: We don't have do it other way. We use makeStyles - personal styles for components. This can be changed separately if needed.

File: map/src/util/Utils.js (hunk 23 above)
Comment (alisa911) [reply map/src/util/Utils.js:88]: `+`

File: map/src/drawer/components/tracks/TracksMenu.jsx (hunk 24 above)
Comment (alisa911) [reply map/src/drawer/components/tracks/TracksMenu.jsx:60]: Cloud and localClientTrack are very different objects with very different fields. We can't work with them same way. For cloud tracks and localServerTrack will be created selectedGpxFile for work with them. I dont change it.

File: map/src/drawer/components/tracks/TracksMenu.jsx (hunk 26 above)
Comment (alisa911) [reply map/src/drawer/components/tracks/TracksMenu.jsx:55]: ?

File: map/src/drawer/util/LocalTrackUtils.js (hunk 27 above)
Comment (alisa911) [reply map/src/drawer/util/LocalTrackUtils.js:1]: `+`
